Update project_usermap creation script (INF-1060)

williamnswanson opened the pull request:
Uses LDAP for determining group membership.

Uses LDAP to determine active users, and only adds a user to the map if they are active (from SOFTWARE-5413)

Still uses the COmanage API to find out which CO groups are projects, but now caches them so if the map needs to be recreated in a short amount of time, it doesn't need to make so many API calls (cache currently stales after 0.5 hours, but changing it should only take changing a global)

williamnswanson committed:
Initial Commit for the new project-usermap script

Uses LDAP search to find active and provisioned groups then compares them to the list project groups in COmanage to determine which LDAP groups and users to build the map out of.

williamnswanson committed:
Remove f string with no placeholders

williamnswanson committed:
Reimplement filtering by members of a group

And clean up of no-longer-used methods

remove collections import

williamnswanson committed:
Add caching for COmanage API data

Writes the osggids to project name dict to a file containing the epoch time the cache was made. If the file is found to exist and was made in the past 0.5 hours (by default), the program will read from the cache instead of making COmanage API calls. Otherwise, the program will get the project data from the API and overwrite any existing cache.

williamnswanson commented:
From talking with Brian L, projects don't need to be provisioned in LDAP, so assuming that LDAP has all the groups that are projects is an error. The project-usermap script should instead use the COmanage API to find the members of project groups (based on the group member entry's Member field being true)

brianhlin commented on osg-comanage-project-usermap.py:
```diff
@@ -1,10 +1,12 @@
+import time
 import getopt
-import collections
+import subprocess
```
I don't think `subprocess` is necessary here as you can use existing `ldap` libraries (see https://github.com/opensciencegrid/topology/blob/master/src/webapp/ldap_data.py#L145-L179).

I'd say generally, you want to use first class libraries instead of `subprocess` callouts as the latter are often pretty brittle. Part of that is that first class libraries will give you Python-native data structures so that you don't have to write hard-to-debug, easy-to-break regular expressions for parsing output.

brianhlin commented on osg-comanage-project-usermap.py:
```diff
@@ -15,7 +17,39 @@
+
+LDAP_AUTH_COMMAND = [
+    "awk", "/ldap_default_authtok/ {print $3}", "/etc/sssd/conf.d/0060_domain_CILOGON.ORG.conf",
+]
```
This assumes a few things that I don't think are necessarily safe to assume:

1. The user running this script can read this file
2. That this file exists

Instead, you should have an option that allows the caller to reference a file containing the password. I prefer that over options specifying the password directly since you don't want the password showing up in plain text in the process tree.

brianhlin commented on osg-comanage-project-usermap.py:
```diff
@@ -15,7 +17,39 @@
+    "-b", "ou=people,o=OSG,o=CO,dc=cilogon,dc=org",
+    "{filter}", "voPersonApplicationUID",
+    "|", "grep", "voPersonApplicationUID",
+    "|", "sort",
```
If you do need to call out to `subprocess`, you certainly don't need pipes. You'll have better output processing tools for you in native Python!

brianhlin reviewed (changes requested):
@williamnswanson this is a solid first attempt but I'm really interested in having you going back to the drawing board using the `ldap3` library because I think it will simplify a lot. You've got a lot of pounds of code here to parse the shell output and I think as a result, it looks like you've got more data structures than you actually need. With the `ldap3` library, I think you can have two data sources:

1. LDAP: All active users that are part of the relevant AP login group. I'd expect the data struct that you get back to include every group that the user is a member of
2. COManage API: All project groups. It looks like you already have this, more or less, but I reserve the right to pick nits in the future :) 

Then project map construction should just be as simple as iterating through the users and taking the intersection of their groups + (2) above

williamnswanson committed:
Name change and add UNIX cluster group creation

williamnswanson committed:
Add project group LDAP provisioning

williamnswanson committed:
project_groups_setup refactoring

williamnswanson committed:
Add ldap search for unprovisioned projects

Also, move get_projects_to_setup() functionality into several different methods for clarity.

williamnswanson committed:
pip install ldap3 inside github CI action

williamnswanson committed:
Created COManage scripts method library

Created a file to hold methods that are, or will likely be, used by multiple scripts in the osg-comanage-rpoject-usermap repo. Refactored existing scripts to use the new library.

williamnswanson committed:
Fix provisioning a project that needed an osggid

Projects that needed an OSGGID identifier weren't being provisioned if a UNIX cluster group already existed. Added identifier updating after adding missing identifiers and moved adding identifiers to before checking what projects need UNIX cluster groups and provisioning.

williamnswanson committed:
Spelling / wording / whitespace changes

No functional changes

williamnswanson committed:
Add more comments and remove COManage data dict

williamnswanson committed:
Initial Commit for the new project-usermap script

Uses LDAP search to find active and provisioned groups then compares them to the list project groups in COmanage to determine which LDAP groups and users to build the map out of.

williamnswanson committed:
Remove f string with no placeholders

williamnswanson committed:
Reimplement filtering by members of a group

And clean up of no-longer-used methods

remove collections import

williamnswanson committed:
Add caching for COmanage API data

Writes the osggids to project name dict to a file containing the epoch time the cache was made. If the file is found to exist and was made in the past 0.5 hours (by default), the program will read from the cache instead of making COmanage API calls. Otherwise, the program will get the project data from the API and overwrite any existing cache.

williamnswanson committed:
Merge branch 'INF-1060.member-removals' of https://github.com/williamnswanson/osg-comanage-project-usermap into INF-1060.member-removals

williamnswanson committed:
project-usermap requested changes
